File: src/main/java/com/spartasystems/holdmail/service/MessageService.java
/*******************************************************************************
 * Copyright 2016 Sparta Systems, Inc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied.
 *
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/

package com.spartasystems.holdmail.service;

import com.spartasystems.holdmail.domain.Message;
import com.spartasystems.holdmail.mapper.MessageListMapper;
import com.spartasystems.holdmail.mapper.MessageMapper;
import com.spartasystems.holdmail.model.MessageList;
import com.spartasystems.holdmail.persistence.MessageEntity;
import com.spartasystems.holdmail.persistence.MessageRepository;
import com.spartasystems.holdmail.smtp.OutgoingMailSender;
import org.apache.commons.lang3.StringUtils;
import org.hibernate.validator.constraints.Email;
import org.hibernate.validator.constraints.NotBlank;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Component;

import javax.validation.constraints.Null;
import java.util.List;

@Component
public class MessageService {

    @Autowired
    private MessageRepository messageRepository;

    @Autowired
    private MessageMapper messageMapper;

    @Autowired
    private OutgoingMailSender outgoingMailSender;

    @Autowired
    private MessageListMapper messageListMapper;

    public Message saveMessage(Message message) {

        MessageEntity entity = messageMapper.fromDomain(message);

        entity = messageRepository.save(entity);

        return messageMapper.toDomain(entity);
    }

    public void deleteMessage(Long id) {
        messageRepository.delete(id);
    }

    public Message getMessage(long messageId) {

        MessageEntity entity = messageRepository.findOne(messageId);

        return messageMapper.toDomain(entity);
    }

    public MessageList findMessages(@Null @Email String recipientEmail, Pageable pageRequest) {

        List<MessageEntity> entities;

        if (StringUtils.isBlank(recipientEmail)) {
            entities = messageRepository.findAllByOrderByReceivedDateDesc(pageRequest);
        }
        else {
            entities = messageRepository.findAllForRecipientOrderByReceivedDateDesc(recipientEmail, pageRequest);
        }

        return messageListMapper.toMessageList(entities);
    }

    public void forwardMessage(long messageId, @NotBlank @Email String recipientEmail) {

        Message message = getMessage(messageId);
        outgoingMailSender.redirectMessage(recipientEmail, message.getRawMessage());
    }

}

File: src/main/resources/static/js/MessageService.js
/*******************************************************************************
 * Copyright 2016 Sparta Systems, Inc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied.
 *
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/

(function () {
    'use strict';

    angular.module('HoldMailApp')

        .service('MessageService', ['$http', function ($http) {

            var MESSAGES_ENDPOINT = '/rest/messages';

            var messageSvc = ({
                messages: [],
                messageDetail: null,
                getMessageList: getMessageList,
                getMessageDetail: getMessageDetail,
                getMessageHTMLURI: getMessageHTMLURI,
                forwardMessage: forwardMessage
            });
            return messageSvc;

            function getMessageHTMLURI(messageId) {
                return MESSAGES_ENDPOINT + '/' + messageId + '/html'
            }

            /**
             * Fetch the list of messages from the backend
             * @param size The max number of messages to return
             * @param page The page number, if the result set is greater than {@link size} messages
             * @param recipientEmail [optional] The email address for the recipient whose emails are being queried
             */
            function getMessageList(size, page, recipientEmail) {

                var params = ['size=' + size, 'page=' + page];
                if (recipientEmail) {
                    params.push('recipient=' + encodeURIComponent(recipientEmail));
                }
                var messageListURI = MESSAGES_ENDPOINT + '?' + params.join('&');

                return $http.get(messageListURI)
                    .then(function (data) {
                        return data;
                    })
                    .catch(function (data, status) {
                        logError('Failed GET to ' + messageListURI, data, status);
                    });
            }

            /**
             * Fetch the full details for a specific message
             * @param messageId The message ID
             */
            function getMessageDetail(messageId) {

                var messageDetailURI = MESSAGES_ENDPOINT + '/' + messageId;
                return $http.get(messageDetailURI)
                    .then(function (data) {
                        return data;
                    })
                    .catch(function (data, status) {
                        logError('Failed GET to ' + messageDetailURI, data, status);
                    });

            }

            /**
             * Forward the specified message to the desired recipient email
             * @param messageId The message ID
             * @param recipientEmail The recipient email
             */
            function forwardMessage(messageId, recipientEmail) {

                var messageForwardURI = MESSAGES_ENDPOINT + '/' + messageId + '/forward';

                return $http.post(messageForwardURI, {
                    recipient: recipientEmail
                })
                    .catch(function (data, status) {
                        logError('Failed POST to ' + messageForwardURI, data, status);
                    });
            }

            function logError(message, data, status) {
                console.log(message + ' [status=' + status + ', message=' + JSON.stringify(data) + ']');

            }

        }]);

}());
